File: solutions/objective-c/bob/1/Bob.m
//
//  Bob.m
//  Bob
//
//  Created by Barrett Clark on 7/1/14.
//  Copyright (c) 2014 Barrett Clark. All rights reserved.
//

#import "Bob.h"

@implementation Bob

- (NSString *) hey:(NSString *)input {
  if ([self isSilent:input]) {
    return @"Fine, be that way.";
  } else if ([self isShouting:input]) {
    return @"Woah, chill out!";
  } else if ([self isAsking:input]) {
    return @"Sure.";
  }
  return @"Whatever.";
}

- (bool) isShouting: (NSString *)input {
  return [[input uppercaseString] isEqualToString:input] &&
          ([input rangeOfCharacterFromSet:[NSCharacterSet letterCharacterSet] options:0].length > 0);
}

- (bool) isAsking: (NSString *)input {
  return [input hasSuffix:@"?"];
}

- (bool) isSilent: (NSString *)input {
  return [[input stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] isEqualToString:@""];
}

@end

File: solutions/elixir/bob/4/bob.exs
defmodule Teenager do
  def hey(message) do
    cond do
      is_nothing(message) -> "Fine. Be that way!"
      is_yelling(message) -> "Woah, chill out!"
      is_asking(message)  -> "Sure."
      true                -> "Whatever."
    end
  end

  defp is_yelling(message), do: String.upcase(message) == message
  defp is_asking(message),  do: String.ends_with?(message, "?")
  defp is_nothing(message), do: String.strip(message) == ""
end